programs/bpf/tests/programs.rs

```diff
@@ -669,3 +672,71 @@ impl InstructionMeter for TestInstructionMeter {
         u64::MAX
     }
 }
+
+#[cfg(any(feature = "bpf_rust"))]
+#[test]
+fn test_program_bpf_instruction_introspection() {
+    solana_logger::setup();
+
+    let GenesisConfigInfo {
+        genesis_config,
+        mint_keypair,
+        ..
+    } = create_genesis_config(50_000);
+    let mut bank = Bank::new(&genesis_config);
+
+    let (name, id, entrypoint) = solana_bpf_loader_program!();
+    bank.add_builtin_loader(&name, id, entrypoint);
+    let bank = Arc::new(bank);
+    let bank_client = BankClient::new_shared(&bank);
+
+    let program_id = load_bpf_program(
+        &bank_client,
+        &bpf_loader::id(),
+        &mint_keypair,
+        "solana_bpf_rust_instruction_introspection",
+    );
+
+    // Passing transaction
+    let account_metas = vec![AccountMeta::new_readonly(
+        solana_sdk::sysvar::instructions::id(),
+        false,
+    )];
+    let instruction = Instruction::new(program_id, &0u8, account_metas);
+    let result = bank_client.send_and_confirm_instruction(&mint_keypair, instruction);
+    println!("result: {:?}", result);
+    assert!(result.is_ok());
+
+    // writable special instructions11111 key, should not be allowed
+    let account_metas = vec![AccountMeta::new(
+        solana_sdk::sysvar::instructions::id(),
+        false,
+    )];
+    let instruction = Instruction::new(program_id, &0u8, account_metas);
+    let result = bank_client.send_and_confirm_instruction(&mint_keypair, instruction);
+    use solana_sdk::transport::TransportError;
+    assert_matches!(
+        result,
+        Err(TransportError::TransactionError(
+            TransactionError::InvalidAccountIndex
+        ))
+    );
+
+    // No accounts, should error
+    let instruction = Instruction::new(program_id, &0u8, vec![]);
+    let result = bank_client.send_and_confirm_instruction(&mint_keypair, instruction);
+    assert!(result.is_err());
+    assert_matches!(
```

Comment:
Just curious, why doesn't `assert_eq!()` work here?

Reply:
```
error[E0369]: binary operation `==` cannot be applied to type `std::result::Result<solana_sdk::signature::Signature, solana_sdk::transpo
rt::TransportError>`
   --> tests/programs.rs:718:5
    |
718 | /     assert_eq!(
719 | |         result,
720 | |         Err(TransportError::TransactionError(
721 | |             TransactionError::InvalidAccountIndex
722 | |         ))
723 | |     );
    | |      ^
    | |      |
    | |______std::result::Result<solana_sdk::signature::Signature, solana_sdk::transport::TransportError>
    |        std::result::Result<_, solana_sdk::transport::TransportError>
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```

Reply:
How about this?
 https://github.com/solana-labs/solana/blob/1211d8118c48df1e2f3e40ebff1734c1fb43d3fc/programs/bpf/tests/programs.rs#L505

Reply:
cool, yea switched to that.